plugins-mu/hm-stack/class-endpoint-controller.php

```diff
@@ -0,0 +1,238 @@
+					'gitData' => [
+						'description'  => esc_html__( 'Data about the current state of git on the site.', 'my-textdomain' ),
+						'type'         => 'object',
+						'context'      => array( 'view' ),
+						'readonly'     => true,
+					],
+				],
+			]
+		] );
+
+		// Fetch all pull requests against this site.
+		register_rest_route( $namespace, 'pull-requests', [
+			'methods'  => \WP_REST_Server::READABLE,
+			'callback' => [ $this, 'get_pull_requests' ],
+			'args'     => [
+				'context' => [
+					'default' => 'view',
+				],
+			],
+			'permission_callback' => [ $this, 'permissions_check' ],
+		] );
+
+		// Fetch average page generation time for this site.
+		register_rest_route( $namespace, 'page-generation', [
+			'methods'  => \WP_REST_Server::READABLE,
+			'callback' => [ $this, 'get_page_generation_time' ],
+			'args'     => [
+				'context' => [
+					'default' => 'view',
+				],
+			],
+			'permission_callback' => [ $this, 'permissions_check' ],
+		] );
+	}
+
+	/**
+	 * Validate whether the current user has appropriate capabilities to fetch HM Stack data or not.
+	 *
+	 * @return bool
+	 */
+	public function permissions_check() : bool {
+		return current_user_can( 'manage_options' );
+	}
+
+	/**
+	 * Make a human-readable error code from various Hm Stack responses.
+	 *
+	 * @param $error WP_Error Error to parse.
+	 * @return WP_Error
+	 */
+	private function get_return_error_message( WP_Error $error ) : WP_Error {
+		switch( $error->get_error_message() ) {
+			case 'Authentication Failed':
+				return new WP_Error(
+					'platform.hmstack.api.could_not_authenticate',
+					'Unable to authenticate to retrieve data.',
+					['status' => WP_Http::INTERNAL_SERVER_ERROR]
+				);
+			default :
+				return new WP_Error(
+					'platform.hmstack.api.could_not_connect',
+					'Unable to connect with the HM Stack API. Error: ' . $error->get_error_message(),
+					['status' => WP_Http::INTERNAL_SERVER_ERROR]
+				);
+		}
+	}
+
+	/**
+	 * @todo:: Need more information to fetch this data.
+	 */
+	public function get_bandwidth_usage() {}
+
+	/**
+	 * Send off environmental data about the site such as gitBranch information and PHP version.
+	 *
+	 * @return WP_REST_Response|\WP_Error
+	 */
+	public function get_environment() {
+		// Check our cache first.
+		$data = wp_cache_get( 'environment', 'hm-stack' );
+		if ( $data !== false ) {
+			return $data;
+		}
+
+		$stack_data = ( new HM_Stack_API() )->get_environment_data();
+
+		// If we've errored out, return a human-friendly message and code.
+		if ( is_wp_error( $stack_data ) ) {
+			// Prevent constant re-fetching in the event of a failure.
+			wp_cache_set( 'activity', $stack_data, 'hm-stack', 5 * \MINUTE_IN_SECONDS );
+			return $this->get_return_error_message( $stack_data );
+		}
+
+		$data = [
+			'environmentData' => [
+				'elasticsearch' => '', // Awaiting this availability
+				'php'           => substr( phpversion(), 0, 5 ),
+				'mySql'         => '', // Will add this later.
```

Comment:
what;s with this casing?

Reply:
This is all going to be feeding into JS exclusively, and this casing is more consistent with React and other JS libraries.

Reply:
On that note, probably could've done `mySQL` or something instead... I don't really care how it's cased as long as it makes sense in the consuming app.

Reply:
The API should really be agnostic, if it's it's only going to be used in JS. Even when working in JS, and calling the WP-API, _most_ I think would expect data returned consistent with the WP API